tarantool/connection_pool.py

```diff
@@ -0,0 +1,466 @@
+        for each connection in the pool. Be careful with reconnect
+        parameters in ConnectionPool since every status refresh is
+        also a request with reconnection. Default is 0.
+        :param float reconnect_delay: Time between reconnect
+        attempts for each connection in the pool. Be careful with
+        reconnect parameters in ConnectionPool since every status
+        refresh is also a request with reconnection. Default is 0.
+        :param StrategyInterface strategy_class: Class for choosing
+        instance based on request mode. By default, round-robin
+        strategy is used.
+        :param int refresh_delay: Minimal time between RW/RO status
+        refreshes.
+        '''
+
+        if not isinstance(addrs, list) or len(addrs) == 0:
+            raise ConfigurationError("addrs must be non-empty list")
+
+        # Verify addresses.
+        for addr in addrs:
+            ok, msg = validate_address(addr)
+            if not ok:
+                raise ConfigurationError(msg)
+        self.addrs = addrs
+
+        # Create connections
+        self.pool = {}
+        self.refresh_delay = refresh_delay
+        self.strategy = strategy_class(self.pool)
+
+        for addr in self.addrs:
+            key = self._make_key(addr)
+            self.pool[key] = PoolUnit(
+                addr=addr,
+                conn=Connection(
+                    host=addr['host'],
+                    port=addr['port'],
+                    user=user,
+                    password=password,
+                    socket_timeout=socket_timeout,
+                    reconnect_max_attempts=reconnect_max_attempts,
+                    reconnect_delay=reconnect_delay,
+                    connect_now=False, # Connect in ConnectionPool.connect()
+                    encoding=encoding,
+                    call_16=call_16,
+                    connection_timeout=connection_timeout)
+            )
+
+        if connect_now:
+            self.connect()
+
+    def __del__(self):
+        self.close()
+
+    def _make_key(self, addr):
+        return '{0}:{1}'.format(addr['host'], addr['port'])
+
+    def _get_new_state(self, unit):
+        conn = unit.conn
+
+        if conn.is_closed():
+            try:
+                conn.connect()
+            except NetworkError as e:
+                msg = "Failed to connect to {0}:{1}".format(
+                    unit.addr['host'], unit.addr['port'])
+                warn(msg, ClusterConnectWarning)
+                return InstanceState(Status.UNHEALTHY)
+
+        try:
+            resp = conn.call('box.info')
+        except Exception as e:
+            msg = "Failed to get box.info for {0}:{1}, reason: {2}".format(
+                unit.addr['host'], unit.addr['port'], repr(e))
+            warn(msg, PoolTolopogyWarning)
+            return InstanceState(Status.UNHEALTHY)
+
+        try:
+            ro = resp.data[0]['ro']
+        except Exception as e:
```

Comment:
Nit: Catching the general `Exception` class is considered as the bad practice. You can catch `(IndexError, KeyError)`.

TBH, I prefer explicit check of value presence (if it is just one value) or validation using a declarative schema. I'm not sure how to better achieve it in Python (there are [many choices](https://stackoverflow.com/q/45812387/1598057)), especially if the goal is to keep a dependency list minimal.

I'm okay to leave it as is. I noted it here to let you look for a better solution out of scope of this pull request if time will permit.

Reply:
Thank you, reworked where is seemed possible.